components/channel_layout/channel_identifier_router/actions.ts

```diff
@@ -155,15 +156,33 @@ export function goToChannelByChannelName(match: Match, history: History): Action
         }
 
         if (!channel || !member) {
+            // Prompt system admin before joining the private channel
+            const user = getCurrentUser(getState());
+            const isSystemAdmin = Utils.isSystemAdmin(user?.roles);
+            if (isSystemAdmin) {
+                const getChannelDispatchResult = await dispatch(getChannelByNameAndTeamName(team, channelName, true));
+                if ('data' in getChannelDispatchResult) {
+                    channel = getChannelDispatchResult.data;
+                    if (channel.type === Constants.PRIVATE_CHANNEL) {
+                        const joinPromptResult = await dispatch(joinPrivateChannelPrompt(teamObj, channel));
+                        if ('data' in joinPromptResult && !joinPromptResult.data.join) {
+                            return {data: undefined};
+                        }
+                    }
+                }
+            }
+
             const joinChannelDispatchResult = await dispatch(joinChannel(getCurrentUserId(state), teamObj!.id, '', channelName));
             if ('error' in joinChannelDispatchResult) {
-                const getChannelDispatchResult = await dispatch(getChannelByNameAndTeamName(team, channelName, true));
-                if ('error' in getChannelDispatchResult || getChannelDispatchResult.data.delete_at === 0) {
-                    await dispatch(fetchMyChannelsAndMembers(teamObj!.id));
-                    handleChannelJoinError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
-                    return {data: undefined};
+                if (!channel) {
+                    const getChannelDispatchResult = await dispatch(getChannelByNameAndTeamName(team, channelName, true));
+                    if ('error' in getChannelDispatchResult || getChannelDispatchResult.data.delete_at === 0) {
```

Comment:
hi! I have a question, is there a reason why you decide to use the -in operator- instead of, in the previous line, destructure the coming values as `const {data, error} = await dispatch(getChannelByName...` ?

Reply:
Yes, I started like that but was getting a ts error. Turns out ActionResult type was like `{data: any} | {error: any}` and doing `{data, error}` makes one type def present and the other absent. So I looked around and saw other code using `in` and I did the same 😅